webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultCsvValidationService.java

```diff
@@ -0,0 +1,200 @@
+package org.opentestsystem.rdw.admin.service.impl;
+
+import org.apache.commons.csv.CSVRecord;
+import org.opentestsystem.rdw.admin.common.security.PermissionScope;
+import org.opentestsystem.rdw.admin.model.CsvSubject;
+import org.opentestsystem.rdw.admin.model.CsvValidationFailure;
+import org.opentestsystem.rdw.admin.model.School;
+import org.opentestsystem.rdw.admin.repository.SchoolRepository;
+import org.opentestsystem.rdw.admin.service.CsvValidationService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Sets.newHashSet;
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+/**
+ * Default implementation of a CsvValidationService
+ */
+@Service
+public class DefaultCsvValidationService implements CsvValidationService {
+
+    private static final String ValidationFormat = "Row: %1$d Failure: %2$s\n";
+    private static final String MoreFailuresFormat = "With %1$d additional validation failures\n";
+    private static final int MaxFailureMessages = 100;
+
+    private final SchoolRepository schoolRepository;
+
+    @Autowired
+    public DefaultCsvValidationService(final SchoolRepository schoolRepository) {
+        this.schoolRepository = schoolRepository;
+    }
+
+    @Override
+    public List<CsvValidationFailure> validateHeaders(final Set<String> csvHeaders) {
+        final List<CsvValidationFailure> failures = new ArrayList<>();
+        final Set<String> headers = csvHeaders.stream().map(String::toLowerCase).collect(Collectors.toSet());
+        if (!headers.containsAll(RequiredHeaders)) {
+            final Set<String> missing = newHashSet(RequiredHeaders);
+            missing.removeAll(headers);
+            failures.add(CsvValidationFailure.builder()
+                    .row(0)
+                    .message("Missing headers " + missing + ", found " + headers)
+                    .build());
+        }
+        return failures;
+    }
+
+    @Override
+    public List<CsvValidationFailure> validateRecords(final Iterator<CSVRecord> recordIterator,
+                                                      final PermissionScope permissionScope) {
```

Comment:
Here I'm accepting a streaming Iterator of CSVRecords.  By requiring that the CSV be ordered, I can validate on a stream and don't need any all-knowing memory constructs that could potentially blow the heap.